### PR TITLE
improve handling of non-finite floating point values

### LINKED ISSUE #480: annotations with +inf, -inf or NaN floats will generate error
Currently, all numeric data is serialized to the web front-end via a JSON serializer.   JSON has no standardized support for +Inf, -Inf or NaN floats.  The Python JSON serializer converts them to special tokens (+Infinity, -Infinity and NaN respectively - note that these are tokens, NOT JSON strings).   The Chrome/V8 JSON deserializer does not recognize these tokens, and throws an error.

end result is that any AnnData containing these special float values in annotations will see a browser error when loading.

Note: this issue affects all AnnData floats (not just annotations).

### PR DESCRIPTION
**Pull request by bkmartinjr**
Several improvements related to issue #480 

* Configure Python JSON encoder to only emit legal/standards compliant JSON. Will throw ValueError it it is handed a non-finite floating point value.
* Catch ValueErrors when serializing to JSON.  These are converted to a 500 HTTP error, and a warning on the CLI.
* add a new CLI option, `--nan-to-num` (named after `numpy.nan_to_num()`), which will make all NaN/Infinity values finite at startup.   This defaults to False as it slows down startup (for large data sets, it can add 10's of seconds to CLI startup).   This option converts to finite using the following rules:
    * NaN converted to zero
    * +Inf converted to the max of the annotation/dataframe value, -Inf to min. 

Fixes #480

### CODE REVIEW

File: server/test/test_scanpy_engine.py [hunk 1]
```diff
@@ -14,7 +14,8 @@
 class UtilTest(unittest.TestCase):
     def setUp(self):
         args = {'layout': 'umap', 'diffexp': 'ttest', 'max_category_items': 100,
-                'obs_names': None, 'var_names': None, 'diffexp_lfc_cutoff': 0.01}
+                'obs_names': None, 'var_names': None, 'diffexp_lfc_cutoff': 0.01,
+                'nan_to_num': True}
```
Comment (csweaver) [inline server/test/test_scanpy_engine.py]: I checked the build time and adding `nan_to_num: True` doesn't increase the time for the test dataset (I was curious)

File: README.md [hunk 2]
```diff
@@ -165,6 +166,12 @@ This is likely because you do not have node and npm installed, we recommend usin
 - `.X` is used to display expression (histograms, scatterplot & colorscale) and to compute differential expression
 - `.obsm` is used for layout
 
+> When I start cellxgene, I get an error `Unexpected HTTP response 500, INTERNAL SERVER ERROR -- Out of range float values are not JSON compliant`, or `Warning: JSON encoding failure - suggest trying --nan-to-num command line option`. What can I do?
```
[the remark this answers is not in the document]
Comment (bkmartinjr) [reply README.md]: yes.

File: server/app/scanpy_engine/scanpy_engine.py [hunk 3]
```diff
@@ -187,40 +187,55 @@ def _IEEE754_special_values_workaround_annotations(self):
...
             for ann in curr_axis:
                 dtype = curr_axis[ann].dtype
                 if dtype.kind == 'f':
-                    not_finite = np.isfinite(curr_axis[ann]) == False   # noqa: E712
-                    if np.count_nonzero(not_finite) > 0:
+                    finite_idx = np.isfinite(curr_axis[ann])
+                    if not finite_idx.all():
+                        curr_axis.loc[np.isnan(curr_axis[ann]), ann] = 0
+                        curr_axis.loc[np.isneginf(curr_axis[ann]), ann] = curr_axis[ann][finite_idx].min()
+                        curr_axis.loc[np.isposinf(curr_axis[ann]), ann] = curr_axis[ann][finite_idx].max()
```
[the remark this answers is not in the document]
Comment (bkmartinjr) [reply server/app/scanpy_engine/scanpy_engine.py]: it was done in consultation with Jeremy, and he believes it is mathematically the right solution given our UX goals.

File: server/app/scanpy_engine/scanpy_engine.py [hunk 4]
```diff
@@ -187,40 +187,55 @@ def _IEEE754_special_values_workaround_annotations(self):
...
-                    "Dataframe X contains floating point NaN or Infinities. "
-                    "These values will be treated as zero."
-                )
-                self._IEEE754_X_warning_issued = True
-
-        return _X
+        # X
+        non_finite_X_found = False
+        if sparse.issparse(self.data._X):
+            coo = self.data._X.tocoo()
```
[the remark this answers is not in the document]
Comment (bkmartinjr) [reply server/app/scanpy_engine/scanpy_engine.py]: It is basically a format where each non-zero element in the matrix is stored separately as a (row, col, data) tuple.    The actual tuples are stored stored in three arrays called row, col and data.    Matrix elements containing zero are omitted.

this format is very convenient from the viewpoint of algorithms which want to process only non-zero elements in the matrix.   If you read the source code for other sparse matrix formats, you will see that they often are implemented by converting to COO format, then doing the particular computation.

File: server/test/test_scanpy_engine.py (hunk 1 above)
Comment (bkmartinjr) [reply server/test/test_scanpy_engine.py]: it does, but the test data set is so tiny that it really isn't something you notice.  Try it on the 10M cell neuron data set for a better test (large X), or on the immune data sets (large annotations).

Comment (bkmartinjr) [reply server/test/test_scanpy_engine.py]: Ah, got it.  Not surprising...

File: server/app/scanpy_engine/scanpy_engine.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (freeman-lab) [reply server/app/scanpy_engine/scanpy_engine.py]: this part looks right to me — you’re absolutely right @csweaver that it risks confounding the data, but i  think it’s better than using a fixed very late constant, which also confounds the data and makes histograms far less useful — so kinda the best of our current options

File: README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (freeman-lab) [reply README.md]: maybe clarify that one error would appear “in the web UI” and the other “in the CLI” (if i have that right)

[the remark this answers is not in the document]
Comment (csweaver) [reply README.md]: Ahh, that is what I didn't understand. The 500 error will only be seen in the developer console.

[the remark this answers is not in the document]
Comment (bkmartinjr) [reply README.md]: Not true - it will be seen in the web UI.

**Comment by freeman-lab** [comment]
looks good to me!

**Pull request merged**